### PR TITLE
Added crafting tweaks functionality to vanillatweaks json method

### PR DESCRIPTION
**Pull request by fleverest**
I altered the error messages to specify whether any error occurred during loading the datapacks or crafting tweaks. Also I removed a heap of trailing whitespace from the README.

### CODE REVIEW

**Review by itzg — COMMENTED** [review]
I know you had good intentions, but I am not happy with all of the whitespaces changes. I need some time to consider how to proceed unless you're willing to reverse those changes.

**Comment by fleverest** [comment]
I don't really care if there's whitespace - it was done by a vim configuration I use which cleans up my code.

Anyway, I reverted the whitespace changes (after 3 attempts, I need coffee).

**Review by itzg — COMMENTED** [review]
Thanks for tightening up the ch ages. The feature looks great. I will also look at the test failure because my recent test script changes might have influenced this to fail without saying much

https://github.com/itzg/docker-minecraft-server/tree/master/tests/setuponlytests/vanillatweaks_file

File: README.md [hunk 1]
```diff
@@ -899,6 +899,12 @@ Example of expected Vanillatweaks file format:
       "coordinates hud"
     ],
     "items": ["armored elytra"]
+  },
+  "craftingtweaks": {
```
Comment (shotah) [inline README.md:903]: Are you able to add this to the test file as well for integration testing?
`./docker-minecraft-server/tests/setuponlytests/forgeapimods_file/example.json`

File: scripts/start-setupVanillaTweaks [hunk 2]
```diff
@@ -33,16 +33,18 @@ if [[ "$VANILLATWEAKS_SHARECODE" ]]; then
...
 if [[ "$VANILLATWEAKS_FILE" ]]; then
   if [ ! -f "$VANILLATWEAKS_FILE" ]; then
     log "ERROR: given VANILLATWEAKS_FILE file does not exist"
     exit 2
   fi
 
   PACKS=$(jq -jc '.packs' $VANILLATWEAKS_FILE)
-  if [ ! "$PACKS" ]; then
-    log "ERROR: unable to retrieve packs from $VANILLATWEAKS_FILE"
+  CRAFTING_TWEAKS=$(jq -jc '.craftingtweaks' $VANILLATWEAKS_FILE)
```
Comment (shotah) [inline scripts/start-setupVanillaTweaks:44]: Makes me want to rename PACKS. :) Thanks for adding this!

File: scripts/start-setupVanillaTweaks [hunk 3]
```diff
@@ -81,4 +83,20 @@ if [[ "$PACKS" ]] && [[ "$VT_VERSION" ]]; then
...
+  if [ ! "$DOWNLOAD_URL" ]; then
+    log "ERROR: unable to retrieve crafting tweak DOWNLOAD_URL from vanillatweaks.net!"
+    exit 2
+  fi
+
+  mkdir -p "$out_dir"
+  if ! get -o $out_dir/crafting_tweaks.zip "https://vanillatweaks.net${DOWNLOAD_URL}"; then
+    log "ERROR: failed to download crafting tweaks from ${DOWNLOAD_URL}"
+    exit 2
+  fi
```
Comment (shotah) [inline scripts/start-setupVanillaTweaks:99]: Looks like cleanup is missing. Probably can be taken out of line 78-83 and made into its own follow up code block.

**Review by shotah — COMMENTED** [review]
Looks great! can we get it added to testing and get the clean up added! Thanks again!

File: scripts/start-setupVanillaTweaks (hunk 2 above)
Comment (itzg) [reply scripts/start-setupVanillaTweaks:44]: FWIW a technique I have done for renaming container variables is like this

https://github.com/itzg/docker-minecraft-server/blob/8a42dfe232e59163654bc0fa536f051210160e16/scripts/start-setupModpack#L190

**Comment by shotah** [comment]
@fleverest I found why this is failing, looks like some real bash commands were added to test.sh. In ci.yml I hard coded on line 39 to use 'sh', that needs to be changed to 'bash'.

file: `.github/workflows/ci.yml`
```bash
run: sh tests/setuponlytests/test.sh -> run: bash tests/setuponlytests/test.sh
```

File: scripts/start-setupVanillaTweaks (hunk 2 above)
Comment (shotah) [reply scripts/start-setupVanillaTweaks:44]: I think this is just my bad choice of names. :) I short handed it, not realizing it was going to get expanded out. :+1:

File: scripts/start-setupVanillaTweaks [hunk 4]
```diff
@@ -81,4 +83,20 @@ if [[ "$PACKS" ]] && [[ "$VT_VERSION" ]]; then
   fi
 fi
 
+# Download and move crafting tweaks
+if [[ "$CRAFTING_TWEAKS" ]] && [[ "$VT_VERSION" ]]; then
```
Comment (shotah) [inline scripts/start-setupVanillaTweaks:87]: After looking at tests, this needs to also be checking for JQ's null string. "null". :(
Something like
```bash
if [[ "$CRAFTING_TWEAKS" ]] && [[ "$CRAFTING_TWEAKS" != "null" ]] && [[ "$VT_VERSION" ]]; then
```

File: scripts/start-setupVanillaTweaks [hunk 5]
```diff
@@ -81,4 +83,20 @@ if [[ "$PACKS" ]] && [[ "$VT_VERSION" ]]; then
...
+  VT_ZIPDATA_URL=https://vanillatweaks.net/assets/server/zipcraftingtweaks.php
+    DOWNLOAD_URL=$(curl -X POST -F "packs=${CRAFTING_TWEAKS}" -F "version=${VT_VERSION}" $VT_ZIPDATA_URL | jq -r '.link')
+  if [ ! "$DOWNLOAD_URL" ]; then
+    log "ERROR: unable to retrieve crafting tweak DOWNLOAD_URL from vanillatweaks.net!"
+    exit 2
+  fi
+
+  mkdir -p "$out_dir"
+  if ! get -o $out_dir/crafting_tweaks.zip "https://vanillatweaks.net${DOWNLOAD_URL}"; then
```
Comment (shotah) [inline scripts/start-setupVanillaTweaks:96]: This is a zip of zips. Does this not need to be unzipped?

File: scripts/start-setupVanillaTweaks (hunk 3 above)
Comment (fleverest) [reply scripts/start-setupVanillaTweaks:99]: No cleanup required, you just move the zip into the datapacks

File: scripts/start-setupVanillaTweaks (hunk 5 above)
Comment (fleverest) [reply scripts/start-setupVanillaTweaks:96]: Works fine locally 🤷 I checked the file downloaded and it's a single zipped directory, unlike the datapacks zip.

File: scripts/start-setupVanillaTweaks (hunk 4 above)
Comment (fleverest) [reply scripts/start-setupVanillaTweaks:87]: I'll add this now

**Comment by fleverest** [comment]
Ok, I think that I addressed everything - let me know if you need anything else from me.

File: scripts/start-setupVanillaTweaks (hunk 4 above)
Comment (fleverest) [reply scripts/start-setupVanillaTweaks]: Added it for the $PACKS variable too.

File: scripts/start-setupVanillaTweaks [hunk 6]
```diff
@@ -54,23 +56,23 @@ if [[ "$VANILLATWEAKS_FILE" ]]; then
 fi
 
 # Download and unzip packs
-if [[ "$PACKS" ]] && [[ "$VT_VERSION" ]]; then
+if [[ "$PACKS" ]] && [[ "$PACKS" != "null"]] && [[ "$VT_VERSION" ]]; then
```
Comment (itzg) [inline scripts/start-setupVanillaTweaks:59]: I'm not sure offhand why test revealed a bash error here. The syntax seems to be defensive. I'll help investigate this too.

Comment (shotah) [reply scripts/start-setupVanillaTweaks:59]: We can also use //empty in the jq like I did for the forge API. Then it'll return empty array like we'd expect. 🤔

File: scripts/start-setupVanillaTweaks (hunk 5 above)
Comment (shotah) [reply scripts/start-setupVanillaTweaks:96]: Agreed, I did end up looking and that is a single zip. My apologies, thought it was like datapacks.

**Comment by shotah** [comment]
@fleverest some of this may be fixed by just fetching from up stream. I'm happy to jump in and get this over the line if you need any help. 

image

File: scripts/start-setupVanillaTweaks (hunk 6 above)
Comment (itzg) [reply scripts/start-setupVanillaTweaks:59]: It just dawned on me that the failing test is the full startup one that only sets EULA and VERSION. On master it's 

https://github.com/itzg/docker-minecraft-server/blob/de0188de3d7bcd919216ed20a04891896c347e6e/tests/fulltests/vanilla-latest/docker-compose.yml#L12

Now I'm wondering why the VANILLATWEAKS logic is even getting hit here.

**Comment by shotah** [comment]
@itzg are we able to close this PR now that we have one coming in for Crafting, Resource, and Datapacks?

**Comment by itzg** [comment]
Sure. #1336 is the new PR and @fleverest I'll make sure to add you as co-author the merge. Fee free to add comments over there too.

**Pull request closed without merging**